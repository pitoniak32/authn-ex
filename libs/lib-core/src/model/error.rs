use derive_more::derive::From;
use serde::Serialize;
use serde_with::{serde_as, DisplayFromStr};

pub type Result<T> = std::result::Result<T, Error>;

#[serde_as]
#[derive(Debug, Serialize, From)]
pub enum Error {
    UserAlreadyExists {
        username: String,
    },
    NonRootUserCantCreate,

    EntityNotFound,

    InvalidObjectIdInserted,

    #[from]
    MongoDb(#[serde_as(as = "DisplayFromStr")] mongodb::error::Error),
}

impl std::fmt::Display for Error {
    fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::result::Result<(), std::fmt::Error> {
        write!(fmt, "{self:?}")
    }
}

impl std::error::Error for Error {}
