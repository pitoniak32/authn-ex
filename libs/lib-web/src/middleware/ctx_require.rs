use std::str::FromStr;

use axum::{
    async_trait,
    body::Body,
    extract::{FromRequestParts, Request, State},
    http::request::Parts,
    middleware::Next,
    response::Response,
};
use axum_extra::extract::{cookie::Cookie, CookieJar};
use base64::prelude::*;
use bson::oid::ObjectId;
use jsonwebtoken::{decode, encode, DecodingKey, EncodingKey, Header, Validation};
use lib_core::{
    ctx::{Ctx, UserInfo},
    model::{user, ModelManager},
};
use rand::distributions::{Alphanumeric, DistString};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use tracing_opentelemetry::OpenTelemetrySpanExt;

use once_cell::sync::Lazy;

use crate::{Error, Result};

pub const ACCESS_TOKEN_KEY: &str = "ACCESS_TOKEN";
pub const REFRESH_TOKEN_KEY: &str = "REFRESH_TOKEN";

pub static KEYS: Lazy<Keys> = Lazy::new(|| {
    let secret = Alphanumeric.sample_string(&mut rand::thread_rng(), 60);
    Keys::new(secret.as_bytes())
});

pub static ACCESS_TOKEN_EXPIRATION: Lazy<chrono::Duration> =
    Lazy::new(|| chrono::Duration::seconds(5));

pub static REFRESH_TOKEN_EXPIRATION: Lazy<chrono::Duration> =
    Lazy::new(|| chrono::Duration::minutes(5));

pub struct Keys {
    pub encoding: EncodingKey,
    pub decoding: DecodingKey,
}

impl Keys {
    pub fn new(secret: &[u8]) -> Self {
        Self {
            encoding: EncodingKey::from_secret(secret),
            decoding: DecodingKey::from_secret(secret),
        }
    }
}

pub fn get_access_token(
    user: &UserInfo,
    encoding_key: &EncodingKey,
    duration: chrono::Duration,
) -> Result<String> {
    let access_token = encode(
        &Header::default(),
        &AccessClaims {
            user: user.clone(),
            exp: (chrono::Utc::now().naive_utc() + duration)
                .and_utc()
                .timestamp() as usize,
        },
        encoding_key,
    )
    .map_err(|_| Error::CtxExt(CtxExtError::CannotEncodeToken))?;

    Ok(access_token)
}

pub fn get_refresh_token(
    user: &UserInfo,
    encoding_key: &EncodingKey,
    duration: chrono::Duration,
) -> Result<String> {
    let refresh_token = encode(
        &Header::default(),
        &RefreshClaims {
            user: user.clone(),
            version: 0,
            exp: (chrono::Utc::now().naive_utc() + duration)
                .and_utc()
                .timestamp() as usize,
        },
        encoding_key,
    )
    .map_err(|_| Error::CtxExt(CtxExtError::CannotEncodeToken))?;

    Ok(refresh_token)
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AccessClaims {
    pub user: UserInfo,
    pub exp: usize,
}

impl FromStr for AccessClaims {
    type Err = Error;

    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
        let parts = s.split(".").collect::<Vec<&str>>();
        let claims = parts
            .get(1)
            .ok_or(Error::CtxExt(CtxExtError::InvalidToken))?;
        let decoded = String::from_utf8(
            BASE64_STANDARD_NO_PAD
                .decode(claims)
                .map_err(|_| Error::CtxExt(CtxExtError::InvalidToken))?,
        )
        .map_err(|_| Error::CtxExt(CtxExtError::InvalidToken))?;
        Ok(serde_json::from_str::<AccessClaims>(&decoded)
            .map_err(|_| Error::CtxExt(CtxExtError::InvalidToken))?)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RefreshClaims {
    pub user: UserInfo,
    pub version: u16,
    pub exp: usize,
}

impl FromStr for RefreshClaims {
    type Err = Error;

    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
        let parts = s.split(".").collect::<Vec<&str>>();
        let claims = parts
            .get(1)
            .ok_or(Error::CtxExt(CtxExtError::InvalidToken))?;
        let decoded = String::from_utf8(
            BASE64_STANDARD_NO_PAD
                .decode(claims)
                .map_err(|_| Error::CtxExt(CtxExtError::InvalidToken))?,
        )
        .map_err(|_| Error::CtxExt(CtxExtError::InvalidToken))?;
        Ok(serde_json::from_str::<RefreshClaims>(&decoded)
            .map_err(|_| Error::CtxExt(CtxExtError::InvalidToken))?)
    }
}

pub async fn ctx_require(ctx: Result<CtxW>, req: Request<Body>, next: Next) -> Result<Response> {
    dbg!(&ctx);
    ctx?;

    Ok(next.run(req).await)
}

// IMPORTANT: This resolver must never fail, but rather capture the potential Auth error and put in in the
//            request extension as CtxExtResult.
//            This way it won't prevent downstream middleware to be executed, and will still capture the error
//            for the appropriate middleware (.e.g., mw_ctx_require which forces successful auth) or handler
//            to get the appropriate information.
pub async fn mw_ctx_resolver(
    State(mm): State<ModelManager>,
    cookies: CookieJar,
    mut req: Request<Body>,
    next: Next,
) -> Response {
    tracing::debug!("{:<12} - mw_ctx_resolve", "MIDDLEWARE");

    let ctx_ext_result = ctx_resolve(mm, &cookies).await;

    if ctx_ext_result.is_err() && !matches!(ctx_ext_result, Err(CtxExtError::TokenNotInCookie)) {
        cookies.remove(ACCESS_TOKEN_KEY);
    }

    // Store the ctx_ext_result in the request extension
    // (for Ctx extractor).
    req.extensions_mut().insert(ctx_ext_result);

    next.run(req).await
}

async fn ctx_resolve(mm: ModelManager, cookies: &CookieJar) -> CtxExtResult {
    let res = match cookies
        .get(ACCESS_TOKEN_KEY)
        .map(|cookie| cookie.value().to_owned())
    {
        Some(token) => {
            // Decode the user data
            let mut validation = Validation::default();
            validation.leeway = 5;
            let decode_result = decode::<AccessClaims>(&token, &KEYS.decoding, &validation);

            dbg!(&decode_result);

            match decode_result {
                Ok(token_data) => {
                    tracing::Span::current()
                        .set_attribute("claims.username", token_data.clone().claims.user.username);
                    Ok(token_data.claims)
                }
                Err(e) if e.kind() == &jsonwebtoken::errors::ErrorKind::ExpiredSignature => {
                    let rf_tok = cookies
                        .get(REFRESH_TOKEN_KEY)
                        .map(|cookie| cookie.value().to_owned())
                        .ok_or(CtxExtError::InvalidToken)?;

                    let token_data = decode::<RefreshClaims>(&rf_tok, &KEYS.decoding, &validation)
                        .map_err(|_| CtxExtError::InvalidToken)?;

                    dbg!(&token_data);

                    tracing::Span::current()
                        .set_attribute("claims.username", token_data.clone().claims.user.username);

                    let new_access_token = get_access_token(
                        &token_data.claims.user,
                        &KEYS.encoding,
                        *ACCESS_TOKEN_EXPIRATION,
                    )
                    .map_err(|_| CtxExtError::InvalidToken)?;

                // TODO: add the new access_token into the cookies.....

                    let _ = cookies
                        .clone()
                        .add(Cookie::new(ACCESS_TOKEN_KEY, new_access_token.clone()));

                    tracing::warn!("created new access_token, using refresh token old one expired");

                    Ok(
                        decode::<AccessClaims>(&new_access_token, &KEYS.decoding, &validation)
                            .map_err(|_| CtxExtError::CannotDecodeToken)?
                            .claims,
                    )
                }
                Err(e) => {
                    tracing::error!(error.kind = ?e.kind(), "encountered error decoding / validating access jwt");

                    Err(CtxExtError::InvalidToken)
                }
            }
        }
        None => Err(CtxExtError::InvalidToken),
    };

    // -- Create CtxExtResult
    match res {
        Ok(claim) => Ctx::new(&claim.user)
            .map(CtxW)
            .map_err(|ex| CtxExtError::CtxCreateFail(ex.to_string())),
        Err(e) => Err(e),
    }
}

// region:    --- Ctx Extractor
#[derive(Debug, Clone)]
pub struct CtxW(pub Ctx);

#[async_trait]
impl<S: Send + Sync> FromRequestParts<S> for CtxW {
    type Rejection = Error;

    async fn from_request_parts(parts: &mut Parts, _state: &S) -> Result<Self> {
        tracing::debug!("{:<12} - Ctx", "EXTRACTOR");
        parts
            .extensions
            .get::<CtxExtResult>()
            .ok_or(Error::CtxExt(CtxExtError::CtxNotInRequestExt))?
            .clone()
            .map_err(Error::CtxExt)
    }
}
// endregion: --- Ctx Extractor

// region:    --- Ctx Extractor Result/Error
type CtxExtResult = core::result::Result<CtxW, CtxExtError>;

#[derive(Clone, Serialize, Debug)]
pub enum CtxExtError {
    TokenNotInCookie,
    InvalidToken,
    CannotEncodeToken,
    CannotDecodeToken,

    UserNotFound,
    ModelAccessError(String),
    FailValidate,
    CannotSetTokenCookie,

    CtxNotInRequestExt,
    CtxCreateFail(String),
}
// endregion: --- Ctx Extractor Result/Error

// #[cfg(test)]
// mod tests {
//     use jsonwebtoken::decode;
//     use jsonwebtoken::Validation;
//     use lib_core::ctx::UserInfo;
//     use rstest::rstest;
//     use uuid::Uuid;

//     #[rstest]
//     #[case(
//         "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyIjp7ImlkIjoiOTk2ZDkwMzUtMzRlNy00YzZmLWE1MjUtOTUyZGI4NDI4ODc2IiwidXNlcm5hbWUiOiJ1bmFtZSIsImRpc3BsYXlOYW1lIjoiVXNlciBOYW1lcnNvbiIsInJlZnJlc2hUb2tlblZlcnNpb24iOjB9LCJleHAiOjE3MjIxOTg2OTJ9.sk8eb1Y70HArDCtHhGXRwtz627I9nuGUU0K0pX5t6uk",
//         UserInfo::with_id(Uuid::parse_str("996d9035-34e7-4c6f-a525-952db8428876").unwrap(), "uname", "User Namerson"),
//     )]
//     fn test_try_from_jwt(#[case] input: &str, #[case] expected: UserInfo) {
//         // Arrange / Act

//         use lib_core::ctx::UserInfo;
//         let result = UserInfo::try_from(input).unwrap();

//         // Assert
//         assert_eq!(result, expected)
//     }

//     #[rstest]
//     #[case(
//         UserInfo::with_id(Uuid::parse_str("996d9035-34e7-4c6f-a525-952db8428876").unwrap(), "uname", "User Namerson"),
//     )]
//     fn test_get_access_token(#[case] input: UserInfo) {
//         // Arrange
//         let mut validation = Validation::default();
//         validation.validate_exp = false;
//         let keys = Keys::new("test_secret_key".as_bytes());

//         // Act
//         insta::assert_debug_snapshot!(decode::<AccessClaims>(
//             &input
//                 .get_access_token(&keys.encoding, chrono::Duration::seconds(1234))
//                 .unwrap(),
//             &keys.decoding,
//             &validation
//         ));
//     }

//     #[rstest]
//     #[case(
//         UserInfo::with_id(Uuid::parse_str("996d9035-34e7-4c6f-a525-952db8428876").unwrap(), "uname", "User Namerson"),
//     )]
//     fn test_get_refresh_token(#[case] input: UserInfo) {
//         // Arrange
//         let mut validation = Validation::default();
//         validation.validate_exp = false;
//         let keys = Keys::new("test_secret_key".as_bytes());

//         // Act
//         insta::assert_debug_snapshot!(decode::<AccessClaims>(
//             &input
//                 .get_refresh_token(&keys.encoding, chrono::Duration::seconds(12345))
//                 .unwrap(),
//             &keys.decoding,
//             &validation
//         ));
//     }
// }
