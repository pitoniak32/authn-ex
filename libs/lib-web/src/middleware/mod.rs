mod ctx_require;

pub use ctx_require::*;
