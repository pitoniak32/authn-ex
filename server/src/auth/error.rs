use axum::{
    http::StatusCode,
    response::{IntoResponse, Response},
    Json,
};
use serde_json::json;

pub type Result<T> = core::result::Result<T, Error>;

#[derive(Debug)]
pub enum Error {
    InvalidToken,
    WrongCredentials,
    TokenCreation,
    MissingCredentials,
    CtxExt,
}

impl IntoResponse for Error {
    fn into_response(self) -> Response {
        let (status, error_message) = match self {
            Error::WrongCredentials => (StatusCode::UNAUTHORIZED, "Wrong credentials"),
            Error::MissingCredentials => (StatusCode::BAD_REQUEST, "Missing credentials"),
            Error::TokenCreation => (StatusCode::INTERNAL_SERVER_ERROR, "Token creation error"),
            Error::InvalidToken => (StatusCode::BAD_REQUEST, "Invalid token"),
            Error::CtxExt => (StatusCode::INTERNAL_SERVER_ERROR, "Failed to retrieve CTX"),
        };
        let body = Json(json!({
            "error": error_message,
        }));
        (status, body).into_response()
    }
}

impl std::fmt::Display for Error {
    fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::result::Result<(), std::fmt::Error> {
        write!(fmt, "{self:?}")
    }
}

impl std::error::Error for Error {}
