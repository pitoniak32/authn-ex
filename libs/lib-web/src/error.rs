use std::sync::Arc;

use axum::{
    http::StatusCode,
    response::{IntoResponse, Response},
};
use derive_more::derive::From;
use serde::Serialize;
use serde_with::serde_as;

use crate::middleware::ctx_require;

pub type Result<T> = std::result::Result<T, Error>;

#[serde_as]
#[derive(Debug, Serialize, From)]
pub enum Error {
    #[from]
    CtxExt(ctx_require::CtxExtError),
}

impl IntoResponse for Error {
    fn into_response(self) -> Response {
        tracing::debug!("{:<12} - model::Error {self:?}", "INTO_RES");

        // Create a placeholder Axum reponse.
        let mut response = StatusCode::INTERNAL_SERVER_ERROR.into_response();

        // Insert the Error into the reponse.
        response.extensions_mut().insert(Arc::new(self));

        response
    }
}
