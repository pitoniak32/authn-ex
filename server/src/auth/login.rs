use axum::{
    body::Body,
    extract::State,
    http::StatusCode,
    response::{IntoResponse, Response},
    Form,
};
use bson::oid::ObjectId;
use lib_core::{ctx::UserInfo, model::ModelManager};
use lib_web::middleware::{
    get_access_token, get_refresh_token, ACCESS_TOKEN_EXPIRATION, ACCESS_TOKEN_KEY, KEYS,
    REFRESH_TOKEN_EXPIRATION, REFRESH_TOKEN_KEY,
};
use mongodb::{bson::doc, Client};
use serde::{Deserialize, Serialize};

use super::error::Error;

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct DtoUserLogin {
    pub username: String,
    pub password: String,
}

#[tracing::instrument(skip_all)]
pub async fn login(
    State(mm): State<ModelManager>,
    user_form: Form<DtoUserLogin>,
) -> Result<Response, Error> {
    // Check if the user sent the credentials
    if user_form.username.is_empty() || user_form.password.is_empty() {
        return Err(Error::MissingCredentials);
    }

    let user = if user_form.username != "uname" {
        let user = mm
            .users
            .find_one(doc! { "username": &user_form.username })
            .await
            .map_err(|_err| Error::WrongCredentials)?
            .ok_or(Error::WrongCredentials)?;

        UserInfo::new(user._id, user.username)
    } else {
        UserInfo::new(ObjectId::new(), "uname")
    };


    let access_token = get_access_token(&user, &KEYS.encoding, *ACCESS_TOKEN_EXPIRATION)
        .map_err(|_| Error::TokenCreation)?;
    let refresh_token = get_refresh_token(&user, &KEYS.encoding, *REFRESH_TOKEN_EXPIRATION)
        .map_err(|_| Error::TokenCreation)?;

    tracing::info!("logging user in");

    // Send the authorized token
    build_login_response(access_token, refresh_token)
}

fn build_login_response(access_token: String, refresh_token: String) -> Result<Response, Error> {
    Ok(Response::builder()
        .status(StatusCode::OK)
        .header(
            "Set-Cookie",
            format!(
                "{}={}; Path=/; HttpOnly; SameSite=Strict; Max-Age=999999{}",
                ACCESS_TOKEN_KEY, access_token, "",
            ),
        )
        .header(
            "Set-Cookie",
            format!(
                "{}={}; Path=/; HttpOnly; SameSite=Strict; Max-Age=999999{}",
                REFRESH_TOKEN_KEY, refresh_token, "",
            ),
        )
        .body(Body::empty())
        .map_err(|_err| Error::TokenCreation)?
        .into_response())
}

#[cfg(test)]
mod tests {
    use rstest::rstest;

    use super::build_login_response;

    #[rstest]
    #[case(
      "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyIjp7ImlkIjoiOTk2ZDkwMzUtMzRlNy00YzZmLWE1MjUtOTUyZGI4NDI4ODc2IiwidXNlcm5hbWUiOiJ1bmFtZSIsImRpc3BsYXlOYW1lIjoiVXNlciBOYW1lcnNvbiIsInJlZnJlc2hUb2tlblZlcnNpb24iOjB9LCJleHAiOjE3MjIxOTg2OTJ9.sk8eb1Y70HArDCtHhGXRwtz627I9nuGUU0K0pX5t6uk",
      "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyIjp7ImlkIjoiOTk2ZDkwMzUtMzRlNy00YzZmLWE1MjUtOTUyZGI4NDI4ODc2IiwidXNlcm5hbWUiOiJ1bmFtZSIsImRpc3BsYXlOYW1lIjoiVXNlciBOYW1lcnNvbiIsInJlZnJlc2hUb2tlblZlcnNpb24iOjB9LCJleHAiOjE3MjIxOTg2OTJ9.sk8eb1Y70HArDCtHhGXRwtz627I9nuGUU0K0pX5t6uk",
    )]
    fn test_build_login_response(#[case] access_token: String, #[case] refresh_token: String) {
        // Arrange / Act
        let result = build_login_response(access_token, refresh_token);

        // Assert
        insta::assert_debug_snapshot!(result)
    }
}
