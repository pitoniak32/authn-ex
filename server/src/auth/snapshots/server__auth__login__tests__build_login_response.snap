---
source: server/src/auth/login.rs
expression: result
---
Ok(
    Response {
        status: 200,
        version: HTTP/1.1,
        headers: {
            "set-cookie": "ACCESS_TOKEN=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyIjp7ImlkIjoiOTk2ZDkwMzUtMzRlNy00YzZmLWE1MjUtOTUyZGI4NDI4ODc2IiwidXNlcm5hbWUiOiJ1bmFtZSIsImRpc3BsYXlOYW1lIjoiVXNlciBOYW1lcnNvbiIsInJlZnJlc2hUb2tlblZlcnNpb24iOjB9LCJleHAiOjE3MjIxOTg2OTJ9.sk8eb1Y70HArDCtHhGXRwtz627I9nuGUU0K0pX5t6uk; Path=/; HttpOnly; SameSite=Strict; Max-Age=999999",
            "set-cookie": "REFRESH_TOKEN=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyIjp7ImlkIjoiOTk2ZDkwMzUtMzRlNy00YzZmLWE1MjUtOTUyZGI4NDI4ODc2IiwidXNlcm5hbWUiOiJ1bmFtZSIsImRpc3BsYXlOYW1lIjoiVXNlciBOYW1lcnNvbiIsInJlZnJlc2hUb2tlblZlcnNpb24iOjB9LCJleHAiOjE3MjIxOTg2OTJ9.sk8eb1Y70HArDCtHhGXRwtz627I9nuGUU0K0pX5t6uk; Path=/; HttpOnly; SameSite=Strict; Max-Age=999999",
        },
        body: Body(
            UnsyncBoxBody,
        ),
    },
)
